pkg/services/pluginsintegration/pluginclient/pluginclient.go

```diff
@@ -0,0 +1,219 @@
+package pluginclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+type PluginReference interface {
+	PluginID() string
+}
+
+type appRef struct {
+	pluginID string
+}
+
+func (r appRef) PluginID() string {
+	return r.pluginID
+}
+
+// AppRef creates a new app PluginReference.
+func AppRef(pluginID string) PluginReference {
+	return &appRef{
+		pluginID: pluginID,
+	}
+}
+
+type DatasourceReference interface {
+	PluginReference
+	DatasourceID() int64
+	DatasourceUID() string
+	Datasource() *datasources.DataSource
+}
+
+type datasourceRef struct {
+	pluginID      string
+	datasourceID  int64
+	datasourceUID string
+	datasource    *datasources.DataSource
+}
+
+func (r datasourceRef) PluginID() string {
+	return r.pluginID
+}
+
+func (r datasourceRef) DatasourceID() int64 {
+	return r.datasourceID
+}
+
+func (r datasourceRef) DatasourceUID() string {
+	return r.datasourceUID
+}
+
+func (r datasourceRef) Datasource() *datasources.DataSource {
+	return r.datasource
+}
+
+type DatasourceReferenceOption func(ref *datasourceRef)
+
+func WithDatasourceID(id int64) DatasourceReferenceOption {
+	return DatasourceReferenceOption(func(ref *datasourceRef) {
+		ref.datasourceID = id
+	})
+}
+
+func WithDatasourceUID(uid string) DatasourceReferenceOption {
+	return DatasourceReferenceOption(func(ref *datasourceRef) {
+		ref.datasourceUID = uid
+	})
+}
+
+func WithDatasource(ds *datasources.DataSource) DatasourceReferenceOption {
+	return DatasourceReferenceOption(func(ref *datasourceRef) {
+		ref.datasource = ds
+
+		if ds != nil {
+			ref.datasourceID = ds.ID
+			ref.datasourceUID = ds.UID
+		}
+	})
+}
+
+// DatasourceRef creates a new datasource PluginReference.
+func DatasourceRef(pluginID string, opts ...DatasourceReferenceOption) PluginReference {
+	ref := &datasourceRef{
+		pluginID: pluginID,
+	}
+
+	for _, opt := range opts {
+		opt(ref)
+	}
+
+	return ref
+}
+
+type QueryDataRequest struct {
```

Comment:
I'm not wild about adding these new request objects -- are they really necessary?

Can we use the existing ones leaving out plugin context?

Reply:
We can, but then I don't really see the point with this proposal any longer. Makes it very confusing for consumers to understand whether they need to provide it or not. This also gives us a place between the interception of grafana and the plugin system to apply common logic.